python/lsst/pipe/tasks/dcrAssembleCoadd.py

```diff
@@ -0,0 +1,781 @@
+        """! Create a DCR-matched template for an exposure.
+
+        @param[in] dcrModels: A list of masked images, each containing
+                              the model for one subfilter.
+        @param[in] visitInfo: lsst.afw.image.VisitInfo for the exposure.
+        @param[in] bbox: Sub-region of the coadd to use for the template.
+        @param[in] wcs: lsst.afw.geom.skyWcs.skyWcs.SkyWcs wcs for the exposure.
+        @param[in] mask: reference mask to use for the template image.
+
+        @return The DCR-matched template, as a maskedImage.
+        """
+        dcrShift = self.dcrShiftCalculate(visitInfo, wcs, self.config.lambdaEff,
+                                          self.config.filterWidth, self.config.dcrNSubbands)
+        templateImage = afwImage.MaskedImageF(bbox)
+        for dcr, model in zip(dcrShift, dcrModels):
+            templateImage += self.convolveDcrModelPlane(model, dcr, bbox=bbox, useFFT=self.config.useFFT)
+        if mask is not None:
+            templateImage.setMask(mask[bbox, afwImage.PARENT])
+        return templateImage
+
+    def dcrResiduals(self, dcrModels, residual, visitInfo, bbox, wcs):
+        """! Prepare a residual image for stacking in each subfilter by applying the reverse DCR shifts.
+
+        @param[in] dcrModels: A list of masked images, each containing
+                              the model for one subfilter.
+        @param[in] residual: The residual masked image for one exposure,
+                             after subtracting the matched template
+        @param[in] visitInfo: lsst.afw.image.VisitInfo for the exposure.
+        @param[in] bbox: Sub-region of the coadd to shift.
+        @param[in] wcs: lsst.afw.geom.skyWcs.skyWcs.SkyWcs wcs for the exposure.
+
+        @return next shifted residual, as a maskedImage.
+        """
+        dcrShift = self.dcrShiftCalculate(visitInfo, wcs, self.config.lambdaEff,
+                                          self.config.filterWidth, self.config.dcrNSubbands)
+        for dcr in dcrShift:
+            yield self.convolveDcrModelPlane(residual, dcr, bbox=bbox,
+                                             useInverse=True, useFFT=self.config.useFFT)
+
+    @staticmethod
+    def calculateRotationAngle(visitInfo, wcs):
+        """! Calculate the sky rotation angle of an exposure.
+
+        @param[in] visitInfo: lsst.afw.image.VisitInfo for the exposure.
+        @param[in] wcs: , wcs for the exposure.
+
+        @return The rotation of the image axis, East from North as a lsst.afw.geom.Angle
+            A rotation angle of 0 degrees is defined with North along the +y axis and East along the +x axis.
+            A rotation angle of 90 degrees is defined with North along the +x axis and East along the -y axis.
+        """
+        parAngle = visitInfo.getBoresightParAngle().asRadians()
+        cd = wcs.getCdMatrix()
+        cdAngle = (numpy.arctan2(-cd[0, 1], cd[0, 0]) + numpy.arctan2(cd[1, 0], cd[1, 1]))/2.
```

Comment:
Can you still do this with `SkyWcs`? If so, that's a bug. We should be hiding the CD matrix from the user. Not sure of the Proper Way to do this.